articles/python-generators/index.md

```diff
@@ -0,0 +1,160 @@
+```
+
+**Generating Fibonacci sequence using generator functions.**
+The Fibonacci Sequence is the series of numbers where next number is found by adding up the two numbers before it.
+0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...
+```python
+def fibonacci_gen():
+    trailing,leading = 0,1
+    while True:
+        yield leading
+        trailing,leading = leading, trailing+leading
+```
+To get the first value of the Fibonacci sequence we generated using generator function above we call the `fibonacci_gen()` function and assign it to variable fib. We then call `next()` on the fib variable, `fib.next()` to get the next number in the sequence.
+```python
+>>>fib = fibonacci_gen()
+>>>fib.next()
+```
+To get the list of the first 10 values for the `fibonacci_gen()` function we call the `fib.next()` inside a for loop with a range of 10
+```python
+for _ in range(10):
+    fib.next()
+```
+
+### Profiling Generator Performance
+```python
+>>> import sys
+# Square root of numbers in a range using list comprehension.
+>>> nums_squared_list = [n * 2 for n in range(10000)]
+>>> sys.getsizeof(nums_squared_list)
+87724
+# Square root of numbers in a range using generator functions.
+>>> nums_squared_generator = (n ** 2 for n in range(10000))
+>>> print(sys.getsizeof(nums_squared_generator))
+125
+```
+List comprehension uses 87724 bytes of memory while generator function uses only 125 bytes of memory. By using generators we save memory as compared to list comprehension where a lot of memory is used.
+
+### Building Generator pipelines
+Generator pipelines allow you to string together code to process large  datasets or streams of data without maxing out your machine’s memory.
+when working with pipelines:-
+- Several pipes can be linked together.
+- Items flow one by one through the entire pipeline.
+- Pipeline functionality can be packaged into a callable function.
+  In order to get the longest name from the text file `names.txt`, We create two generator methods.
+  - The first generator method full_names yields names line by line stripping off the new line from each line.
+  - The second generator method lengths yields a tuple with name and an integer representing its length.
+  - To get the longest name we use max and pass in the lengths generator object and pass a key for the lengths.
+```python
+def get_longest_name():
+	full_names = (name.strip() for name in open("text/names.txt"))
+	lengths = ((name,len(name)) for name in full_names)
+	longest = max(lengths,key=lambda x:x[1])
+```
+### Conclusions 
+Now you better understand python generators. They are useful when dealing with complex data sets and performing memory-intensive tasks. You can find more on python generators from the python [wiki](https://wiki.python.org/moin/Generators/). These will help you create better applications with cleaner code and offer alternatives to creating nested for loops and working with list comprehensions which are memory intensive
```

Comment:
Now you better understand python generators->We understand the concepts of generators in python in this article.